psl/parser-database/src/coerce_expression.rs

```diff
@@ -14,6 +14,25 @@ macro_rules! impl_coercions {
                 coerce::<$lt>(super::coerce_opt::$name, $expected_type)(expr, diagnostics)
             }
             )*
+
+            pub fn function_or_constant_with_span<'a>(
+                expr: &'a ast::Expression,
+                diagnostics: &mut Diagnostics,
+            ) -> Option<(&'a str, &'a [ast::Argument], ast::Span)> {
+                match super::coerce_opt::function_or_constant_with_span(expr) {
+                    Some(val) => Some(val),
+                    None => {
+                        diagnostics.push_error(DatamodelError::new_type_mismatch_error(
+                            "constant or function",
+                            expr.describe_value_type(),
+                            &expr.to_string(),
+                            expr.span(),
+                        ));
+
+                        None
+                    }
+                }
+            }
```

Comment:
I'm wondering if there wouldn't be a more general solution instead of adding that to the macro definiton.

Reply:
I don't really like this macro. It makes us skip boilerplate, until it doesn't fit anymore and makes adding new coercing functions a bit of a drag.